PR: Add the possibility to rename IPython consoles

tjolive opened the pull request:
Fixes #1962

----

**_Tabs_** and **_TabBar_** classes (spyder/widgets/tabs.py) accept now a parameter (_rename_tabs_) to enable the tab renaming - defaults no disable but **_IPythonConsole_** instantiates **_Tabs_** with _rename_tabs_=True. If activated, the **_TabBar_** creates a new object **_EditTabNamePopup_** which is triggered by right double clicking on a tab.

The **_EditTabNamePopup_** instantiates _QLineEdit_ and places it on top of the tab when activated. While activated, if the user clicks outside the _QLineEdit_ of presses the ESC key, the changes are discarded (i.e. _QLineEdit_ gets invisible).

Structure: IPythonConsole > Tabs > TabBar > EditTabNamePopUp

Some details that could be further improved:
- The tab names have a margin of 9. This number was reached by trial&error, could not find it anywhere on the _QTabBar_/_QTabWidget_ config. It works but would like to get this value dynamically
- The _QLineEdit_ defaults to a white background so I didn't include any config to change this. Not sure if we would like to see this customizable
- The **_IPythonConsole_** class (spyder/plugins/ipythonconsole.py) already has a _rename_client_tab_ function which is not used. Not sure if we want to keep it

goanpeca commented:
@tjolive awesome, thanks for the help!

Could you paste some screenshots of the process or maybe an animation? Maybe use something like http://www.cockos.com/licecap/

tjolive commented:
here you are:
![tab_rename](https://cloud.githubusercontent.com/assets/25354277/22556479/93fd8cf8-e967-11e6-974c-a30c3b696766.gif)

Realized now that I cannot see any cursor when editing but my windows is behaving strangely...

Coded and tested on:
- Windows 7
- Spyder 3.2.0.dev0
- Python 3.5.1 32bits
- Qt 5.7.1
- PyQt5 5.7.1

goanpeca commented on spyder/plugins/ipythonconsole.py:
```diff
@@ -1509,3 +1515,8 @@ def _create_client_for_kernel(self, connection_file, hostname, sshkey,
         self.register_client(client)
+
+    def tab_name_editor(self):
+        """Triggers the tab name editor"""
```
Lets add a `.` at the end a docstring is a sentence.

tjolive replied:
ok

tjolive replied:
ok

goanpeca commented on spyder/widgets/tabs.py:
```diff
@@ -48,6 +48,12 @@ def __init__(self, parent, ancestor):
 
+        # Tab name editor
+        self.rename_tabs = rename_tabs
+        if self.rename_tabs is True:
```
`self.rename_tabs is True` this is redundant, we can just use

`if self.rename_tabs:`

tjolive replied:
Usually I use this as an implicit "assert isinstance(rename_tabs, bool)" but I'll align with your code style

goanpeca commented on spyder/widgets/tabs.py:
```diff
@@ -48,6 +48,12 @@ def __init__(self, parent, ancestor):
+        self.rename_tabs = rename_tabs
+        if self.rename_tabs is True:
+            # Creates tab name editor
+            self.tab_name_editor = EditTabNamePopup(self)
```
This variable is only getting created if true, so maybe we should define a 

`self.tab_name_editor = None` above

tjolive replied:
Good idea, missed that

goanpeca commented on spyder/widgets/tabs.py:
```diff
@@ -111,7 +117,20 @@ def dropEvent(self, event):
-        
+
+    def mouseDoubleClickEvent(self, event):
+        """Override Qt method to trigge the tab name editor"""
```
`.` at the end

tjolive replied:
ok

goanpeca commented on spyder/widgets/tabs.py:
```diff
@@ -111,7 +117,20 @@ def dropEvent(self, event):
+                # Tab is valid, call tab name editor
+                self.tab_name_editor.edit_tab(index)
+        else:
+            # Even is not interessant, raise to parent
```
Event ? missed a t?

tjolive replied:
typo, thanks for spotting

goanpeca commented on spyder/widgets/tabs.py:
```diff
@@ -337,3 +357,85 @@ def move_tab_from_another_tabwidget(self, tabwidget_from,
+
+
+class EditTabNamePopup(QLineEdit):
+    """Popup on top of the tab to edit its name"""
```
`.` at the end

tjolive replied:
ok

goanpeca commented on spyder/widgets/tabs.py:
```diff
@@ -337,3 +357,85 @@ def move_tab_from_another_tabwidget(self, tabwidget_from,
+    """Popup on top of the tab to edit its name"""
+
+    def __init__(self, parent):
+        QLineEdit.__init__(self, parent=None)
```
We can add the same docstring here

tjolive replied:
ok

goanpeca commented on spyder/widgets/tabs.py:
```diff
@@ -337,3 +357,85 @@ def move_tab_from_another_tabwidget(self, tabwidget_from,
+        self.setFrame(False)
+
+        # Align with tab name
+        self.setTextMargins(9, 0, 0, 0)
```
We will need to check how this looks on other OS

tjolive replied:
How are these cross-platform tests usually performed?

rlaverde replied:
It looks nice in Linux :smile:

rlaverde replied:
It looks nice on Linux :smile:

tjolive replied:
Can you post some video? On OSX works also (see below comment).

But I can't see the cursor, will fix it.

goanpeca commented on spyder/widgets/tabs.py:
```diff
@@ -337,3 +357,85 @@ def move_tab_from_another_tabwidget(self, tabwidget_from,
+        self.setTextMargins(9, 0, 0, 0)
+
+        # Track with tab is being edited
+        self.tab_index = None
```
It would be good to have a similar structure to how other widgets work

```

# Instance variables

# Widgets

# Widget setup

# Layouts

# Signals and slots

```

tjolive replied:
do you have any template for such documentation?

goanpeca replied:
Nope, but most of the widgets follow that convention

```python

class SomeWidget():
    # Variables
    self.some_variable = None
  
    # Widgets
    self.super_widget = SomeSuperWidget()

    # Widget setup
    self.super_widget.setSomething(True)

    # Layouts
    layout = QVBoxLayout()
    layout.addWidget(self.super_widget)
    self.setLayout(layout)

   # Signals and slots
   self.super_widget.sig_some_signal.connect(self.some_other_method)

rlaverde replied:
@goanpeca We need to document that somewhere.

rlaverde replied:
I added to the wiki https://github.com/spyder-ide/spyder/wiki/Dev:-Coding-Style#widget-structure

tjolive replied:
Will align the code with this.

Took the liberty to add to the wiki the docstring format as mentioned here:
https://github.com/spyder-ide/spyder/wiki/Dev:-Coding-Style#docstrings

goanpeca commented on spyder/widgets/tabs.py:
```diff
@@ -337,3 +357,85 @@ def move_tab_from_another_tabwidget(self, tabwidget_from,
+        self.tab_index = None
+
+    def eventFilter(self, widget, event):
+        """Catches clicks outside the object and ESC key press"""
```
`.` at the end

tjolive replied:
ok

rlaverde replied:
Catch (infinitive form)

tjolive replied:
ok

goanpeca commented on spyder/widgets/tabs.py:
```diff
@@ -337,3 +357,85 @@ def move_tab_from_another_tabwidget(self, tabwidget_from,
+        return QLineEdit.eventFilter(self, widget, event)
+
+    def edit_tab(self, index):
+        """Activates the edit tab"""
```
`.` at the end

tjolive replied:
ok

rlaverde replied:
Activate (infinitive form)

tjolive replied:
ok

tjolive commented:
Tested on OSX Sierra 10.12.2 (16C67, Python 2.7.13 64bits, Qt 5.6.2, PyQt5 5.6 on Darwin

![on_osx](https://cloud.githubusercontent.com/assets/25354277/22585108/6473c7d8-e9f6-11e6-9f82-c86f1ba8134e.gif)

When editing, the text is not as aligned as in windows, maybe because the close icon is rendered on the left. In any way it looks nice and intuitive.

The cursor is also missing in OSX, did you have the same behaviour on linux?

tjolive commented on spyder/plugins/ipythonconsole.py:
```diff
@@ -756,6 +756,10 @@ def get_plugin_actions(self):
                _("Open a new IPython console connected to an existing kernel"),
                triggered=self.create_client_for_kernel)
         
+        rename_tab_action = create_action(self, _("&Rename tab"),
```
ok

goanpeca commented:
This is looking great

tjolive commented:
Hi guys,

new commit addresses all feedback (I hope) and adds a blinking cursor, which I felt was missing.

But now the Travis CI is failing on something that I didn't change and was working on the first commit.

```
+for f in 'spyder/*/*/*.py'
+[[ spyder/utils/introspection/docstrings.py == *test*/test_* ]]
+[[ spyder/utils/introspection/docstrings.py == spyder/external/*/*.py ]]
+[[ spyder/utils/introspection/docstrings.py == spyder/utils/external/*.py ]]
+[[ spyder/utils/introspection/docstrings.py == spyder/utils/help/*.py ]]
+[[ spyder/utils/introspection/docstrings.py == spyder/utils/ipython/start_kernel.py ]]
+[[ spyder/utils/introspection/docstrings.py == spyder/utils/ipython/spyder_kernel.py ]]
+[[ spyder/utils/introspection/docstrings.py == spyder/utils/site/sitecustomize.py ]]
+[[ spyder/utils/introspection/docstrings.py == spyder/utils/introspection/plugin_client.py ]]
+[[ spyder/utils/introspection/docstrings.py == spyder/widgets/externalshell/systemshell.py ]]
+[[ spyder/utils/introspection/docstrings.py == spyder/widgets/ipythonconsole/__init__.py ]]
+python spyder/utils/introspection/docstrings.py
Traceback (most recent call last):
  File "spyder/utils/introspection/docstrings.py", line 26, in <module>
    from jedi.evaluate.iterable import Array, FakeSequence, AlreadyEvaluated
ImportError: cannot import name 'Array'
The command "./continuous_integration/travis/modules_test.sh" exited with 1.
Done. Your build exited with 1.

```

Any ideas?

Do you need any more changes on this PR?

Cheers,
Telmo

ccordoba12 commented:
Please merge with the 3.x branch to fix that error.

goanpeca commented:
@tjolive I think you made a mistake :-|, cause now the PR is $^%^

tjolive commented:
Argh, I see, I tried to do via the github desktop app and it defaulted to master. Reverting and merging with 3.x...

goanpeca commented:
If you are not sure how to fix this @tjolive, just start from scratch

tjolive commented:
Revert was ok. Merged with 3.x but it didn't trigger a re-check...

Will try something else, otherwise I'll issue a new PR

goanpeca commented:
Its not ok...

<img width="613" alt="screen shot 2017-02-22 at 8 45 15 am" src="https://cloud.githubusercontent.com/assets/3627835/23214032/42b56504-f8db-11e6-9971-17f245ad40a9.png">

ccordoba12 commented:
@tjolive, please remove your merge commit with

     git reset --hard HEAD~1

(maybe also your revert commit too with that command), then push with

    git push -f tjolive PR1962_rename_ipython_consoles

tjolive commented:
thanks @ccordoba12 for the help, who would have thought that coding was the easier part...

After reseting, I get a message mentioning master and not 3.x:

```
git reset --hard HEAD~1
HEAD is now at cfdfe242 Merge remote-tracking branch 'refs/remotes/origin/master' into PR1962_rename_ipython_consoles
```

and can't push
```
C:\Script\51_dev\spyder>git push -f tjolive PR1962_rename_ipython_consoles
fatal: 'tjolive' does not appear to be a git repository
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
```

will try to update from 3.x on the desktop gui and sync, if that doesn't work I'll issue a new clean PR

goanpeca commented:
`git push -f origin PR1962_rename_ipython_consoles`

ccordoba12 commented:
> git push -f origin PR1962_rename_ipython_consoles

Yeah, silly me!! I was thinking in my own work flow, which is adding user names as remote's in my clone :-p

tjolive commented:
still not working... :(

Last try...

tjolive commented:
Guys, I cannot fix it, no matter what I do with it it still hangs all those commits (presumably from the master) and the checks keep on failing.

If you're ok with it, you can delete this PR and I'll issue a new one.

Sorry for this mess...

rlaverde commented:
@tjolive I reverted the wrong merge, and merge it with the branch 3.x, I hope it could be merged now.

It was necessary to reset to the last commit hash:
`git reset --hard 2595f68`

tjolive commented:
@rlaverde thank you very much for this effort, now it looks way better :)

...but the continuous-integration/travis-ci/pr keeps on failing (now at a different stage):
```
+for f in 'spyder/*/*/*.py'
+[[ spyder/widgets/variableexplorer/dataframeeditor.py == *test*/test_* ]]
+[[ spyder/widgets/variableexplorer/dataframeeditor.py == spyder/external/*/*.py ]]
+[[ spyder/widgets/variableexplorer/dataframeeditor.py == spyder/utils/external/*.py ]]
+[[ spyder/widgets/variableexplorer/dataframeeditor.py == spyder/utils/help/*.py ]]
+[[ spyder/widgets/variableexplorer/dataframeeditor.py == spyder/utils/ipython/start_kernel.py ]]
+[[ spyder/widgets/variableexplorer/dataframeeditor.py == spyder/utils/ipython/spyder_kernel.py ]]
+[[ spyder/widgets/variableexplorer/dataframeeditor.py == spyder/utils/site/sitecustomize.py ]]
+[[ spyder/widgets/variableexplorer/dataframeeditor.py == spyder/utils/introspection/plugin_client.py ]]
+[[ spyder/widgets/variableexplorer/dataframeeditor.py == spyder/widgets/externalshell/systemshell.py ]]
+[[ spyder/widgets/variableexplorer/dataframeeditor.py == spyder/widgets/ipythonconsole/__init__.py ]]
+python spyder/widgets/variableexplorer/dataframeeditor.py
./continuous_integration/travis/modules_test.sh: line 64:  3883 Segmentation fault      (core dumped) python "$f"

```

ccordoba12 commented:
> but the continuous-integration/travis-ci/pr keeps on failing (now at a different stage)

Don't worry about that, sometimes it fails for unknown reasons.

gaw89 commented:
@tjolive, are you planning to come back to this PR at some point to see if it can get pushed through?  I would be very interested in having this feature available in Spyder.

tjolive commented:
@gaw89: I have this as closed on my end, the coding part is done and
tested, now it's up to the owner to integrate this.

On 29 June 2017 at 19:34, gaw89 <notifications@github.com> wrote:

> @tjolive <https://github.com/tjolive>, are you planning to come back to
> this PR at some point to see if it can get pushed through? I would be very
> interested in having this feature available in Spyder.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/spyder-ide/spyder/pull/4092#issuecomment-312038043>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AYLgJTYgdqwzAdNg6vKSS4gMfMBgymg5ks5sI-BDgaJpZM4L03CB>
> .
>

gaw89 commented:
@tjolive, I see.  I asked you because of the conflicting files listed in the PR status.  Is that your responsibility to figure out the conflict?

@ccordoba12, what needs to be done in order to get this PR accepted?  If I can help, I'm happy to do so.

tjolive commented:
@gaw89: i see the conflict now, i don't recall having it when i concluded
the coding some months ago. Let me take a look and revert.

Anyway,  @ccordoba12 <https://github.com/ccordoba12> can we have your view
on this one?


On 30 Jun 2017 14:33, "gaw89" <notifications@github.com> wrote:

@tjolive <https://github.com/tjolive>, I see. I asked you because of the
conflicting files listed in the PR status. Is that your responsibility to
figure out the conflict?

@ccordoba12 <https://github.com/ccordoba12>, what needs to be done in order
to get this PR accepted? If I can help, I'm happy to do so.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/spyder-ide/spyder/pull/4092#issuecomment-312255230>,
or mute
the thread
<https://github.com/notifications/unsubscribe-auth/AYLgJWagatd4i2flLc8YMmJEI_htm8lIks5sJOr_gaJpZM4L03CB>
.

gaw89 commented:
@tjolive, I see.  Perhaps the conflict cropped up since you submitted the PR.  If there's anything I can do to help, let me know.  This feature would be incredibly useful for me and a great addition to an already great IDE!

ccordoba12 commented:
@tjolive, don't worry about the conflict. We need to fix a couple of things before merging your work, but we'll keep working on your branch.

ccordoba12 commented:
@gaw89, this will be part of Spyder 3.2.

gaw89 commented:
@ccordoba12, awesome!  I'll keep an eye out for it.  Is there a release date for 3.2?

ccordoba12 commented:
Hopefully in a couple of weeks.

peendebak commented:
@tjolive @ccordoba12 Is there an option to rename the tabs from the command line itself? E.g. `import spyder; spyder.tags.rename('my name');`

ccordoba12 commented:
No, there isn't.

peendebak commented:
@ccordoba12 Is it technically possible? E.g. can the spyder gui elements be controlled from the ipython console? If so, how, then I can make a PR to create this functionality.
